### ISSUE TITLE
Numerous fails in LandIce post merge 

### ISSUE THREAD

**Issue by ikalash** [SNLComputation/Albany#525]
Looks like there are numerous fails in LandIce after you merged in your branch @bartgol .  Did you run the LandIce tests prior to merging it in?  Can you please look at the issue / fix it?

http://cdash.sandia.gov/CDash-2-3-0/viewTest.php?onlyfailed&buildid=89549

**Comment by bartgol** [comment]
On it.

**Comment by ikalash** [comment]
Thanks @bartgol !

**Comment by bartgol** [comment]
I fixed some, and got a plan for some other, but I have an E3SM meeting today and tomorrow. I will try to find time to do this.

I am also having a hard time figuring out why the `SteadyHeatConstrainedOpt2D_Dirichlet_Mixed_Params_Analysis_Epetra` test hangs. It hangs in Piro, when it tries to print the final solution... I'm at a loss on that one.

**Comment by ikalash** [comment]
Thanks Luca.  Maybe we should revert the problematic commit and create a new PR for the change / fixes given that the fix is nontrivial?  Are you opposed to that?

**Comment by bartgol** [comment]
Turns out fixing the bugs I already planned to fix, magically fixed the hanging Analysis test. I won't investigate, just look up and say "thanks" to the Developer above, and be happy.

**Comment by bartgol** [comment]
@ikalash I pushed another commit, which fixes all base+landlice tests on my workstation. Let's see what happens tonight.

**Comment by ikalash** [comment]
Seems ATO fails with the change on my workstation - did it pass for you?

**Comment by ikalash** [comment]
By "with the change", I meant the original fix, not the debug one.

**Comment by bartgol** [comment]
I haven't tested other physics. I had limited time due to the meeting today, and wanted to get the fix in before tonight. I will inspect in my spare time tomorrow. I'll look at the nightlies output.

**Comment by ikalash** [comment]
In addition to the ATO tests failing there's a new ALI failure on mayer: 

https://my.cdash.org/viewTest.php?onlyfailed&buildid=1712144

Also, it appears that all the CISM-Albany tests hang, and I suspect you'd have something similar for MALI.  I actually had to make some changes to the CALI/MALI interfaces to get the code to compile (https://github.com/SNLComputation/Albany/commit/ac64c723fdd0568d4d858f0f61e0f824c64d3b52 ).  Can you please have a look at it?  Do some of the changes you made yesterday to fix some of the hanging need to be propagated into the CALI/MALI interface code?

**Comment by bartgol** [comment]
It's possible. I need to check.

**Comment by bartgol** [comment]
I fixed the ATO tests. I don't see anything wrong with the CISM/MPAS interfaces. Where is the cdash output of those tests?

**Comment by ikalash** [comment]
Thanks, Luca.  There are definitely still some issues.  In addition to the CISM tests failing (hanging), I'm finding that the epetra version of a performance test I'm writing hangs as well: https://my.cdash.org/testDetails.php?test=50412166&build=1711708 .  The CISM tests run on Cori (https://my.cdash.org/viewSubProjects.php?project=Albany&date=2019-10-07) but b/c of the nature of how the tests failed, the output did not get posted to the CDash site.  You can find the output is attached and can be found here on Cori: /project/projectdirs/piscees/nightlyCoriCDash .  I'd be surprised if you didn't see similar behavior with MALI tests.  I can send you the configure / input files for the problematic tests.  It should be very easy to reproduce on any given platform using the scripts I have.

[nightly_log_coriCismAlbanyRun.txt](https://github.com/SNLComputation/Albany/files/3704916/nightly_log_coriCismAlbanyRun.txt)

**Comment by ikalash** [comment]
Is what I did here correct?  https://github.com/SNLComputation/Albany/commit/ac64c723fdd0568d4d858f0f61e0f824c64d3b52

**Comment by bartgol** [comment]
> Is what I did here correct? ac64c72
Yes, that's correct.

I'm really not sure, the two interfaces seem correct. There may be something wrong inside some stk class, that only gets exposed by those tests. Have you tried to run your perf test locally, with 2 ranks, in gdb, interrupt the exec when you think it's hanging, and see what's the deadlock about? That'd be the fastest way to see where the issue is. And it might be it's the same that hangs the cism tests. Your test had a very large mesh, so this is clearly not doable with 2 ranks, but perhaps the issue shows up again with a very coarse mesh and O(2) ranks...

**Comment by ikalash** [comment]
Could you try running MALI tests to see if they work?

**Comment by ikalash** [comment]
No, I haven't run my case locally.  Would you be able to look at it?  Ironically the test that is hanging right now was the target production problem that had the slow-down in the finite element assembly in the Gather Extruded 2D field evaluator.  Now it's taking infinitly longer than before b/c it hangs... 

The data for the case can be found here on blake:/home/projects/albany/nightlyCDashAlbanyBlake/aliPerfTests/nightlyCDash/build/ALIPerfTestsBuild/perf_tests/ant-2-20km .  The relevant case is given by input_albany_ml_LineSmoothing.yaml .

**Comment by bartgol** [comment]
> Could you try running MALI tests to see if they work?
I can try. I never built MALI (sigh), so I may need you to point me to some script/docs.

I will try to run the hanging ali testcase locally on my machine. Might be that fixing that will fix MALI too...

**Comment by mperego** [comment]
@bartgol, I'll do that

**Comment by mperego** [comment]
I mean, I'll try running MALI

**Comment by bartgol** [comment]
@ikalash I cannot run such a huge mesh on 2 ranks, unfortunately. Is the mesh "morally" the same as the one we use in the FO_AIS tests (i.e., a finer mesh, but with the same sidesets/nodesets)? If so, I can try to run your input file with that mesh, and see if it hangs. If it doesn't, we're out of luck, since I can't see anything wrong with the code, and debugging a problem that size is prohibitive...

**Comment by ikalash** [comment]
@mperego : is the ant-2-20km mesh you gave me some time ago for ML vs. MueLu performance tests morally similar to the FO_AIS tests in the repo?

**Comment by ikalash** [comment]
It's worth seeing if we can figure out what is happening in MALI and CALI in case it's related.  It's a bit problematic for me to test CALI now b/c I have it only on Cori which will get doing down soon.  I am going to try to build it on one of my workstation.  

If the large tests hangs we should probably revert all the "speedups", since that was the target problem and now it does not run.  Anyway, lets see what happens with MALI.

**Comment by bartgol** [comment]
I don't think we should revert the changes; "does not run" and "very very slow" are equally bad. We should just focus on fixing whatever is making the tests hang.

**Comment by mperego** [comment]
> @mperego : is the ant-2-20km mesh you gave me some time ago for ML vs. MueLu performance tests morally similar to the FO_AIS tests in the repo?

I think so. It's worth trying the the FO_AIS input file on a FO_GIS problem.

**Comment by bartgol** [comment]
> It's worth trying the the ant-2-20km input file on a FO_AIS or FO_GIS problem.

Good idea. That would make us understand whether it's a problem size issue or a parameter-list options issue.

**Comment by ikalash** [comment]
May be related: on mayer, FO_GIS_CoupledThickness_Epetra is failing now (after the merge that broke things): https://my.cdash.org/testDetails.php?test=50437557&build=1712705 .  It hangs similarly to the AIS ML problem that hangs.

**Comment by ikalash** [comment]
I've done a bit of digging and here is some info on the CISM tests, in case the symptoms spark some ideas on what could be wrong:

- 2 of the tests hang: ross_diagnostic_Epetra, confined-shelf_diagnostic_Epetra.  Note that they are both Epetra tests - the Tpetra analogs of these tests do not fail.
- The tests run fine prior to the merging of the PR that is being discussed here (sha 13fec1d6c200e15983d69dd571e093766efc50c9 of Albany is good).
- The issue happens on > 1 procs.
- Sometimes the problem manifests itself with the following MPI error:
```
7: [mockba:52396] *** An error occurred in MPI_Bcast
7: [mockba:52396] *** reported by process [601227265,2]
7: [mockba:52396] *** on communicator MPI COMMUNICATOR 3 CREATE FROM 0
7: [mockba:52396] *** MPI_ERR_TRUNCATE: message truncated
7: [mockba:52396] *** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
7: [mockba:52396] ***    and potentially your MPI job)
7: [mockba.ca.sandia.gov:52392] 1 more process has sent help message help-mpi-errors.txt / mpi_errors_are_fatal
7: [mockba.ca.sandia.gov:52392] Set MCA parameter "orte_base_help_aggregate" to 0 to see all help / error messages
```
- The issue is with this line: https://github.com/SNLComputation/Albany/blob/master/src/LandIce/interface_with_cism/ali_driver.cpp#L528 .  I don't know why that's an issue now, but it was definitely https://github.com/SNLComputation/Albany/commit/77fb33cf49c88d65d222e58d2f89c6411a5abdf2 that broke things.

**Comment by bartgol** [comment]
That line is definitely a red herring, since it does not depend on anything that those commits changed. I'm still going through the code, trying to figure out what's wrong.

**Comment by bartgol** [comment]
Ah. The FO_GIS_Coupled_Thickness_Epetra test hangs in a combine operations inside `Application::computeGlobalResidualImpl`. More precisely, the 4 ranks are all stuck _on the same_ `MPI_Barrier` call. This seems unrelated to the changes recently introduced.

I am starting to suspect that the recent commits simply unveiled an issue that was there before, but for some reason we got away with so far.

Are the hangings happening only on mayer? If so, we may speculate it is related to the MPI installation. Yes, it didn't happened before, but the fact that it gets stuck in an MPI_Barrier is very suspicious. Was there an update of mayer modules?

Also, I used clang on mayer (like @ikalash ). I might try with gcc, and see if something changes.

**Comment by ikalash** [comment]
Could you check if the same thing is happening for the CALI tests that fail?  I can give you an account on my machine / scripts to build it there.  The test is very small.  You were right about the red herring.

I have a trip next week that I need to prepare for so I don't have a ton of time to dig into this, and I don't want to leave it hanging indefinitely - this is why I'm asking if you could help.

Within Albany, the hang for that particular test happens only on mayer, yes.  But we see hangs on other problems - e.g. the AIS bigger problem, and CALI - those happen everywhere.  I'm curious if it's the same issue across the board.  CALI is small enough to debug on unlike the AIS problem.  Since those issues happen across the board, it's not gcc specific.

**Comment by ikalash** [comment]
Also, what is the line where the code is hanging with the ```MPI_Barrier``` call?

**Comment by bartgol** [comment]
I ran into this interesting (very old) thread on openmpi forums.

[](https://users.open-mpi.narkive.com/P9LyJqMV/ompi-users-processes-stuck-in-mpi-barrier)

The user had a different setup/configuration, but his issue seems akin to ours, considering that the FO_GIS_Coupled_Thickness_Epetra test hangs with 4 ranks, but not with 3 or 5.

**Comment by ikalash** [comment]
Ok, interesting.  Maybe the issue is different then for CALI.  I get the hang or MPI error on any number of procs > 1.

**Comment by bartgol** [comment]
> Also, what is the line where the code is hanging with the `MPI_Barrier` call?

Buried deep in some Epetra communication related file in the stack of an `Export` call.

**Comment by bartgol** [comment]
@ikalash sure, if you get me an account, I can get in and try to dig more.

**Comment by ikalash** [comment]
That would be great, thanks, @bartgol !  I will email you instructions for how to logon / build the code later today.

**Comment by ikalash** [comment]
Regarding:
> 
> Buried deep in some Epetra communication related file in the stack of an `Export` call.

But to what line in Albany does the backtrace lead to?  You said it's in computeGlobalResidual, but what line?

**Comment by ikalash** [comment]
> I have a trip next week that I need to prepare for so I don't have a ton of time to dig into this, and I don't want to leave it hanging indefinitely - this is why I'm asking if you could help.

No pun intended :).

**Comment by bartgol** [comment]
> Regarding:
> 
> > Buried deep in some Epetra communication related file in the stack of an `Export` call.
> 
> But to what line in Albany does the backtrace lead to? You said it's in computeGlobalResidual, but what line?

Oh, sorry. It's [this](https://github.com/SNLComputation/Albany/blob/7e44e3f13f7ed63ae0db0245f5f01520b3a44294/src/Albany_Application.cpp#L1289) line.

**Comment by ikalash** [comment]
Hmmm, that's not particularly helpful...  it's just an export that is underlying the combine...  very puzzling.  I just sent you instructions for how to build CALI on my workstation @bartgol , as well as given you an account.

**Comment by bartgol** [comment]
I ran the ross test in a debugger, with 4 ranks. The test hanged for a bit during the creation of the mesh and then crashed. Setting some breakpoints, I found out the issue is in the bulk data final setup (the `modification_ends` call). At some point, the processes perform a reduction to verify if there has been an error in any of the other processes, so they do the following:

```
all_reduce(parallel(), ReduceMax<1>(&error_flag));
ThrowErrorMsgIf(error_flag, error_msg.str());
```
On entry, all processors have `error_flag=0`, but on exit, ranks 1,2, and 3 have `error_flag=32767`, which happens to be the largest value for a signed short int (although `error_flag` is declared as `int`), while rank 0 still has `error_flag=0`. So rank 0 is hanging, waiting for the other ones, which have thrown an exception and are trying to crash. When I run it with 16 ranks (not through the debugger), it hangs for a long time (I don't know if then it crashes, cause I killed it after some time). I assume that the error signal does not reach the ORTE and the program does not crash.

Bottom line: something is off, and I don't know why. I will try running the program through a memory corruption checker, to see if something comes out. But this becomes more and more a mystery every day.

**Comment by bartgol** [comment]
Ok, it seems as the four ranks are not "fully linked". There's this sequence of calls at some point:

```
  const int result_reduce =
    MPI_Reduce(arg_in,arg_out,arg_len,MPI_BYTE,mpi_op,0,arg_comm);

  const int result_bcast =
    MPI_Bcast(arg_out,arg_len,MPI_BYTE,0,arg_comm);
```

I'm not sure how the MPI_Reduce call works, namely, whether non-root ranks (here the root is 0) should be waiting at this call until all procs get there, or if they simply bcast their value to root and continue. However, I _do_ know (hope?) that the second line, the bcast, _should_ be blocking for all ranks, which wait there until the root rank (here, 0) is done communicating them the specified buffer. Instead, all non-root ranks simply carry on, as if the bcast happened, while I am instead holding back the root rank (that is, I 'continue' execution past the bcast on non-root ranks, while root rank is still at the MPI_Reduce call).

So my only guess is that we (or, at least, someone) is messing up the communicators. I printed them, and on all ranks the comm seems legit (size 4, with ranks from 0 to 3, and same context id), but I still suspect something is off with them. At least, I can figure out any other reason why the non-root ranks would not stop at the bcast call, waiting for rank 0 to actually bcast.

Note: I checked the bcast call, and the arg_len argument is consistent (i.e., 8 bytes, 4 for the int, and one for padding in the WorkType structure that stk uses for reduction).

Note 2: I am not sure this is what makes the tests hangs. Perhaps this only shows up in debug builds, but it's a red herring. I'm going to comment out those lines in Trilinos, and see what happens.

**Comment by ikalash** [comment]
Thanks for digging into this @bartgol .  Would it be worthwhile to look at the code prior to merging your PR, as the issue was not happening then?  Might shed some light onto what is different.

**Comment by ikalash** [comment]
Regarding the test hanging: it happens in the nightlies, which are release, so it's not just in the debug builds.

**Comment by bartgol** [comment]
Ok, I put some printf statements in one of the mesh checks in BulkData.cpp. This is the code snippet:


```
    if (parallel_size() > 1) {
      printf("ok on proc %d is %d\n",parallel_rank(),verified_ok);
      all_reduce( parallel() , ReduceMin<1>( & verified_ok ) );
      printf("on rank %d the reduce min of verified_ok across %d ranks gives %d\n",parallel_rank(),parallel_size(),verified_ok);
    }   
```
The original code did not obviously include the printf statements. This is in the routine `BulkData::comm_mesh_verify_parallel_consistency`. The output I get is the following:

```
ok on proc 0 is 1
ok on proc 1 is 1
on rank 1, the reduce min of verified_ok across 4 ranks gives 32767
[...]
ok on proc 2 is 1
on rank 2, the reduce min of verified_ok across 4 ranks gives 32765
[...]
ok on proc 3 is 1
on rank 3, the reduce min of verified_ok across 4 ranks gives 32764
on rank 0, the reduce min of verified_ok across 4 ranks gives 1
```

The `[...]` symbol means that there is some other output in between (it's just the exception being thrown on most ranks). Two comments:

1) why is the result of the reduce op printed _before_ all ranks print their local result? Yes, output could be flushed at random times, but the calls to the first printf should _all_ happen before the calls to the second print (since there are bcast inside the all_reduce call, which should act like barriers), and I believe printf flushes every time it finds a new line arg (the `\n` in my format string), so _all ranks_ should print the first line _before any rank_ prints the second. The fact that they don't, means something is corrupted in the comms groups. At least I hope, cause that's the easiest explanation...

2) the result is correct on rank 0, but not on other ranks.

**Comment by bartgol** [comment]
Note: the reduction works fine in all other albany tests, so the issue must be related to what we do in the cism mesh (i.e., there's no bug in trilinos...I think...).

**Comment by bartgol** [comment]
> Thanks for digging into this @bartgol . Would it be worthwhile to look at the code prior to merging your PR, as the issue was not happening then? Might shed some light onto what is different.

I'm rebuilding the previous albany now.

**Comment by bartgol** [comment]
Last update of the day (will continue tomorrow).

Resetting the CISM mesh struct files to what they were before the merge (and consequently re-adding the thyra utils needed, such as `getLocalElement`) fixes the issue. All the rest of Albany _still_ uses the new indexer class, so the issue is confined to the cism mesh.

Best guess so far (to be checked tomorrow): something goes wrong when retrieving the Epetra objects form the Thyra ones (recall that the ThyraEpetraAdapters are not as nice as the Tpetra ones, and they use rcp's extra_data to keep track of stuff).

**Comment by bartgol** [comment]
I think I fixed the CISM tests. I have to wait till tomorrow to see the nightlies, but at the very least, the fix I did was needed.

The issue (which must be checked in the rest of Albany) is that the `createGlobalLocalIndexer(vs)` is a *collective* operations with respect to the processes that are in the communicator associated with the underlying map (to the vs). In the CISM mesh struct construction, the indexer was created inside if region that were skipped by some processes.

**Comment by ikalash** [comment]
Thanks for fixing this @bartgol !  How come this was not an issue before?

I was hoping the CISM issue would resolve the Albany hangs, but it sounds like it will not...

**Comment by bartgol** [comment]
Sorry, I realized I didn't answer fully.

The other hanging tests in Albany are probably due to the same reason (not all ranks being involved in the creation of the indexer). I have to check, but that's my suspicion. Where are they hanging though? I only see an LCM test failing, due to python issues (although it fails across the board, so it might be an issue in our runtest scripts).

**Comment by ikalash** [comment]
I'm overseas so I can't check the srn builds.  Wasn't there a land ice test hangings on Mayer?  Then there is the largish performance ais test that started this issue.  You can see that one on the son cdash under blake.

**Comment by bartgol** [comment]
d728b67 seems to have fixed the FO_Thermo_Wet_Bet test for me on mayer. At this point I'm hopeful the commit may fix all the tests that started failing after the indexer branch went in. If tomorrow there's still something broken, I'll take another look, otherwise I'll close this issue.

**Comment by bartgol** [comment]
I'm going to close this issue, since the remaining failing tests are LCM tests (likely failing for other reasons), and one LandIce test on Mayer (also likely for other reasons).

**Comment by ikalash** [comment]
Thanks for fixing it!  It looks like the performance test is back up, so I think we’re good!
